sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIOStorageReadTest.java

```diff
@@ -1351,4 +1353,20 @@ public void testReadFromBigQueryIOWithTrimmedSchema() throws Exception {
 
     p.run();
   }
+
+  private static org.apache.arrow.vector.types.pojo.Field field(
```

Comment:
Should there be an test with the Arrow format in this file? It looks like it's still only testing Avro, and these new methods aren't used.

Reply:
Ah it looks like these methods are only used in `BigQueryStorageReaderTest`. That file doesn't have any end-to-end tests of reading Arrow either though.

Could you add a test like `testReadFromBigQueryIO` that uses the Arrow format?

Reply:
Hi @TheNeuralBit I’m writing the test for that case, for that I need to create a `ReadRowsResponse` backed by an `ArrowRecordBatch`, however I can't find a way to create an instance of `ArrowRecordBatch` with data, would you by any chance know where I can find an example of that? Thanks

Reply:
Yeah the Arrow Java API isn't super intuitive. You should be able to follow the code in `ArrowConversionTest` to create a `VectorSchemaRoot`, then you can use a [`VectorUnloader`](https://arrow.apache.org/docs/java/reference/index.html?org/apache/arrow/vector/VectorUnloader.html) to make it into an `ArrowRecordBatch`.

Reply:
Hi @TheNeuralBit I’m still having some problems with this implementation, I followed the examples you gave me, which helped me understand more the Arrow side, but when I was implementing it I found that `ReadRowsResponse` uses a different `ArrowRecordBatch` implementation found in `com.google.cloud.bigquery.storage.v1.ArrowRecordBatch` and with that now I have two questions:
- In order to convert from `org.apache.arrow.vector.ipc.message.ArrowRecordBatch` to `com.google.cloud.bigquery.storage.v1.ArrowRecordBatch`, from what I can see I would need to write the `org.apache.arrow.vector.ipc.message.ArrowRecordBatch` to an InputStream and then [parse](https://googleapis.dev/java/google-cloud-bigquerystorage/1.8.3/com/google/cloud/bigquery/storage/v1/ArrowRecordBatch.html#parseFrom-com.google.protobuf.CodedInputStream-) [it](https://googleapis.dev/java/google-cloud-bigquerystorage/1.8.3/com/google/cloud/bigquery/storage/v1/ArrowRecordBatch.html#parseFrom-java.io.InputStream-) to `com.google.cloud.bigquery.storage.v1.ArrowRecordBatch`, is that correct? and is there a specific coder needed to do that? I found the `Avro` implementation uses a `GenericDatumWriter`. Is there any other more straightforward way to do that conversion?
- Is it correct for the Arrow [code](https://github.com/apache/beam/blob/ed63ed19c67c7c0a1ce7c2888c44fb1238419c32/sdks/java/extensions/arrow/src/main/java/org/apache/beam/sdk/extensions/arrow/ArrowConversion.java) to use `org.apache.arrow.vector.ipc.message.ArrowRecordBatch` instead of `com.google.cloud.bigquery.storage.v1.ArrowRecordBatch`?
Thanks

Reply:
Oh shoot sorry about that, I completely missed that there is a separate BigQuery `ArrowRecordBatch`. I think you're proposed approach is mostly correct, but rather than parsing the serialized record batch, you'll want to make a builder and set the appropriate values, (e.g. with [setSerializedRecordBatch](https://googleapis.dev/java/google-cloud-bigquerystorage/1.8.3/com/google/cloud/bigquery/storage/v1/ArrowRecordBatch.Builder.html#setSerializedRecordBatch-com.google.protobuf.ByteString-)):

```java
ArrowRecordBatch bigqueryBatch = ArrowRecordBatch.newBuilder()
    .setRowCount(..)
    .setSerializedRecordBatch(serializedBytes)
    .build()
```

> Is it correct for the Arrow code to use `org.apache.arrow.vector.ipc.message.ArrowRecordBatch` instead of `com.google.cloud.bigquery.storage.v1.ArrowRecordBatch`?

Yes this is correct. The bigquery ArrowRecordBatch is only relevant for BigQueryIO, while `ArrowConversion` should be more general purpose (there may be other IOs that produce arrow data in the future).

Reply:
Hi, to I'm currently checking [BigQueryStorageSourceBase](https://github.com/apache/beam/blob/fea337953f492aeaf05792690fb05257d16915b9/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryStorageSourceBase.java), after dealing with some errors writing the Arrow test I realized that this class is always trying to read as AVRO format, I will upload the changes soon for revision.

Reply:
`ArrowConversion` can convert an Arrow schema to a Beam schema. You should probably follow the pattern you're using for the data (Arrow to Beam rows to Avro records to BigQuery TableRows):
- Arrow schema Beam Schema (using ArrowConversion)
- Beam Schema to `AvroSchema` (using AvroUtils)
- `AvroSchema` to BigQuery `TableSchema` (using  BigQueryAvroUtils)

Reply:
Hi @TheNeuralBit I'm having some problems writing the tests for the Arrow Read, during deserialization of the `ArrowRecordBatch` I'm getting an error `Expected RecordBatch but header was 0` ```
at org.apache.arrow.vector.ipc.message.MessageSerializer.deserializeRecordBatch(MessageSerializer.java:360)
	at org.apache.beam.sdk.extensions.arrow.ArrowConversion.rowFromSerializedRecordBatch(ArrowConversion.java:260)
```
I think the problem is related at how I serialize the ArrowRecordBatch in the test or how it is getting deserialized in the ArrowConversion, the error specifically makes me believe that the format is getting lost when converting to and from the `RecordBatch` in the Arrow library and the one in the bigquery library, but I'm really not that Knowledgeable of Arrow so I'm getting a little lost finding a solution.

Reply:
Here is a commit that fixes the serialization issue: https://github.com/TheNeuralBit/beam/commit/b4a7c092715b9299025cc2f094c0b07895a545be. It looks like there are other issues though. Not sure if it's a problem with the test data or the implementation,

Reply:
Thanks, just checked it out, it seems that the new problem is that when converting to an Avro `GenericRecord` in the ArrowReader the values are `null`. Im doing some digging to see why this may be, it appears that when deserializing from the `com.google.cloud.bigquery.storage.v1.ArrowRecordBatch` found in the `ReadSession` to `org.apache.arrow.vector.ipc.message.ArrowRecordBatch` the values are already null.

Reply:
@TheNeuralBit Also, I made a small check immediately attempting to deserialize in the same way ArrowConversion is doing it, and I'm getting the same null [there](https://github.com/MiguelAnzoWizeline/beam/commit/19d160116948fe1e2206a18a9d7b931dfc44fead)
Data is being set to the Record [appropriately](https://github.com/MiguelAnzoWizeline/beam/blob/19d160116948fe1e2206a18a9d7b931dfc44fead/sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIOStorageReadTest.java#L604) following the examples I saw in ArrowConversion.
It makes me think that deserialization process is wrong somehow.

Reply:
I pushed another commit that addresses this: https://github.com/apache/beam/commit/0533fb5029fdad8e9209a51fa56a393d84c160bc

The most important part was switching to [using](https://github.com/apache/beam/commit/0533fb5029fdad8e9209a51fa56a393d84c160bc#diff-6e7aa8820ce67ec29c07471edad242ff4bb0b2eabf05286d6d35d0e4a434ac4dR644) `VectorSchemaRoot.allocateNew` instead of allocating each vector independently as was suggested [here](https://github.com/apache/beam/pull/10384/files#r372106201).

After doing that I found the test failed because some arrow objects were not getting closed, so I proposed some changes that manage the lifetimes of the arrow objects (e.g. switching to `RecordBatchRowIterator extends Iterator<Row>, AutoCloseable`)